File: NERP.Admin.Web/ClientApp/src/app/civilworks/services/civilwork.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { HttpClient, HttpParams } from "@angular/common/http";

import { Civilwork } from '../models/civilwork';
import { PaginationRequest, SortOrder } from '../../common/models/paginationRequest';
import { PaginationResponse } from '../../common/models/paginationResponse';
import { ServiceResponse } from '../../common/models/serviceResponse';

@Injectable()
export class CivilworkService {

    getCivilworksApi = 'Civilworks/GetCivilworks/';
    getCivilworkApi = 'Civilworks/GetCivilwork/';
    checkNameExistsApi = 'Civilworks/CheckNameExists/';
    saveCivilworkApi = 'Civilworks/SaveCivilwork/';
    deleteCivilworkApi = 'Civilworks/DeleteCivilwork/';

    constructor(private http: HttpClient) { }

    getCivilworks(paginationRequest?: PaginationRequest): Observable<PaginationResponse> {

        if (paginationRequest == null)
            paginationRequest = {
                page: null,
                pageSize: null,
                filter: null,
                order: {
                    orderByProperty: "name",
                    sortOrder: SortOrder.descending
                }
            };

        return this.http.post<PaginationResponse>(this.getCivilworksApi, paginationRequest);
    }

    getCivilwork(civilworkId): Observable<ServiceResponse> {

        const params = new HttpParams()
            .set('civilworkId', civilworkId);

        return this.http.get<ServiceResponse>(this.getCivilworkApi, { params });
    }

    async checkNameExists(name: string): Promise<ServiceResponse> {

        const params = new HttpParams()
            .set('name', name);

        return await this.http.get<ServiceResponse>(this.checkNameExistsApi, { params }).toPromise();
    }

    saveCivilwork(civilwork: Civilwork): Observable<ServiceResponse> {

        return this.http.post<ServiceResponse>(this.saveCivilworkApi, civilwork);
    }

    deleteCivilwork(civilworkId: string): Observable<ServiceResponse> {

        const params = new HttpParams()
            .set('civilworkId', civilworkId);

        return this.http.delete<ServiceResponse>(this.deleteCivilworkApi, { params });
    }
}

File: NERP.NetworkEntities/CivilworkService.cs
using Common.Config;
using Common.DataAccess.Models;
using Common.DataAccess.Mongo;
using Microsoft.Extensions.Options;
using NERP.NetworkEntities.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;

namespace NERP.NetworkEntities
{
    public class CivilworkService : ICivilworkService
    {
        private readonly IRepository<Civilwork> _civilworksRepository;
        private const string CivilworksRepository = "MST_Civilworks";

        public CivilworkService(IOptions<DbConfig> dbConfig)
        {
            var databaseConfig = dbConfig.Value;

            var dbSettings = new DbSettings { ConnectionString = databaseConfig.ConnectionString, Database = databaseConfig.Database };

            _civilworksRepository = new MongoRepository<Civilwork>(dbSettings, CivilworksRepository);
        }

        public async Task<Civilwork> GetCivilwork(string civilworkId)
        {
            if (string.IsNullOrEmpty(civilworkId)) return null;

            return await _civilworksRepository.GetByIdAsync(civilworkId);
        }

        public async Task<IEnumerable<Civilwork>> GetCivilworks(SearchFilter searchFilter, string orderBy, SortOrder sortOrder, int? page, int? pageSize)
        {
            Expression<Func<Civilwork, bool>> filterExpression = x => true;

            if (searchFilter != null && searchFilter.Filters.Count > 0)
                filterExpression = ExpressionBuilder.GetExpression<Civilwork>(searchFilter);

            IList<Civilwork> civilworks = null;

            if (page != null && pageSize != null)
                civilworks = (await _civilworksRepository.GetPaginatedAsync(filterExpression, orderBy, sortOrder, (int)page, (int)pageSize)).ToList();
            else
                civilworks = (await _civilworksRepository.GetAllAsync(filterExpression, null, orderBy, sortOrder)).ToList();

            return civilworks;
        }

        public async Task<long> GetCivilworksCount(SearchFilter searchFilter = null)
        {
            Expression<Func<Civilwork, bool>> filterExpression = x => true;

            if (searchFilter != null && searchFilter.Filters.Count > 0)
            {
                filterExpression = ExpressionBuilder.GetExpression<Civilwork>(searchFilter);

                //var filter = new Filter<Civilwork>();

                //foreach (var item in searchFilter.Filters)
                //{
                //    filter.By(item.PropertyName, Operation.Contains, item.Value, Connector.And);
                //}

                //filter.By("Name", Operation.Contains, " John");

                //filterExpression = new FilterBuilder().GetExpression<Civilwork>(filter);
            }

            return await _civilworksRepository.CountAsync(filterExpression);
        }

        public async Task<string> CheckNameExists(string name)
        {
            var civilwork = await _civilworksRepository.GetOneAsync(x => x.Name == name);

            return civilwork?.Id;
        }

        public async Task<Civilwork> SaveCivilwork(Civilwork civilwork)
        {
            return await _civilworksRepository.UpdateOneAsync(civilwork);
        }

        public async Task<bool> DeleteCivilwork(string civilworkId)
        {
            return (await _civilworksRepository.DeleteByIdAsync(civilworkId)) == 1;
        }

    }
}
